adam_core/dynamics/impacts.py

```diff
@@ -0,0 +1,277 @@
+
+            # Create futures
+            futures = []
+            idx = np.arange(0, len(orbits))
+            for idx_chunk in _iterate_chunks(idx, chunk_size):
+                futures.append(
+                    impact_worker_ray.remote(
+                        idx_chunk, orbits_ref, self.__class__, num_days
+                    )
+                )
+
+            # Get results as they finish (we sort later)
+            unfinished = futures
+            while unfinished:
+                finished, unfinished = ray.wait(unfinished, num_returns=1)
+                (propagated, impacts) = ray.get(finished[0])
+                propagated_list.append(propagated)
+                impact_list.append(impacts)
+
+            propagated = qv.concatenate(propagated_list)
+            impacts = qv.concatenate(impact_list)
+
+        else:
+            propagated, impacts = self._detect_impacts(orbits, num_days)
+
+        return propagated, impacts
+
+
+def calculate_impacts(
+    orbits: Orbits,
+    num_days: int,
+    propagator: Propagator,
+    num_samples: int = 1000,
+    processes: Optional[int] = None,
+    seed: Optional[int] = None,
+):
+    """
+    Calculate the impacts for each variant orbit generated from the input orbits.
+
+    Parameters
+    ----------
+    orbits : `~adam_core.orbits.orbits.Orbits`
+        Orbits for which to calculate impact probabilities.
+    num_days : int
+        Number of days to propagate the orbits.
+    propagator : `~adam_core.propagator.propagator.Propagator`
+        Propagator to use for orbit propagation.
+    num_samples : int, optional
+        Number of samples to take over the period, by default 1000.
+    processes : int, optional
+        Number of processes to use for parallelization, by default all available.
+    seed : int, optional
+        Seed for random number generation, by default None.
+
+    Returns
+    -------
+
+    """
+    assert issubclass(
+        propagator.__class__, ImpactMixin
+    ), "Propagator must support impact detection."
+    logger.info("Generating variants...")
+    variants = VariantOrbits.create(
+        orbits, method="monte-carlo", num_samples=num_samples, seed=seed
+    )
+    logger.info("Detecting impacts...")
+    results, impacts = propagator.detect_impacts(
+        variants,
+        num_days,
+        max_processes=processes,
+    )
+
+    return results, impacts
+
+
+def calculate_impact_probabilities(variants, impacts):
+    """
+    Calculate the impact probabilities for each variant orbit generated from the input orbits.
+    Parameters
+    ----------
+    variants : `~adam_core.orbits.variants.VariantOrbits`
+        Variant orbits for which to calculate impact probabilities.
+    impacts : `~adam_core.orbits.impacts.Impacts`
+        Impacts for the variant orbits.
+    Returns
+    -------
+    impact_probabilities : `~adam_core.orbits.impact_probabilities.ImpactProbabilities`
+        Impact probabilities for the variant orbits.
```

Comment:
Nice, I like it.

Reply:
I ended up not including the min/max time, due to complexity. If we want the table to include probabilities for objects with no impacts, those would need to be nullable. Although an `.as_column` can be nullable, I was running into issues with merging attributes again when trying to pass `[None]`. 

It seemed better to include the variants with no impacts, because it's nice to see explicitly that they have no impacts. I think since it's really really easy to get this data with the linkage, we just use that if we want min/max. Most of the time we probably want groupings of impacts by day or month.